android/greengrass-nucleus-android/src/main/java/com/aws/greengrass/android/service/DefaultGreengrassComponentService.java

```diff
@@ -0,0 +1,375 @@
+                    workerExitCode = EXIT_CODE_FAILED;
+                    return;
+                }
+            }
+
+            try {
+                // Enter critical section to avoid sudden interruption of Nucleus startup process
+                enterCriticalSection();
+                logger.atDebug().log("Entered critical section for Nucleus startup");
+
+                AndroidPlatform platform = (AndroidPlatform) Platform.getInstance();
+                platform.setAndroidAPIs(DefaultGreengrassComponentService.this, packageManager, componentManager);
+
+                ProvisionManager provisionManager = BaseProvisionManager.getInstance(getFilesDir());
+                final String[] nucleusArguments = provisionManager.prepareArguments();
+                kernel = GreengrassSetup.main(nucleusArguments);
+
+                // Clear system properties
+                provisionManager.clearSystemProperties();
+
+                if (!provisionManager.isProvisioned()) {
+                    provisionManager.writeConfig(kernel);
+                }
+
+                // Startup is done, now we can leave critical section and allow to terminate Nucleus
+                leaveCriticalSection();
+                logger.atDebug().log("Nucleus startup is done, left critical section");
+
+                // waiting for Thread.interrupt() call
+                join();
+            } catch (InterruptedException e) {
+                logger.atInfo().kv("workerExitCode", workerExitCode)
+                        .log("Nucleus worker thread interrupted");
+            }  catch (Throwable e) {
+                logger.atError().setCause(e)
+                        .log("Error while running Nucleus core worker thread");
+            } finally {
+                // Ensure we are out of critical section before exiting
+                leaveCriticalSection();
+                if (kernel != null) {
+                    kernel.shutdown();
+                    kernel = null;
+                }
+            }
+        }
+    }
+
+    @Override
+    protected void onExitRequested() {
+        // Do not stop service immediately, just ask worker thread to exit
+        // After that the service will be stopped by kernel.shutdown() call
+        componentWorkerThread.stopWorker();
+    }
+
+    @Override
+    protected void onServiceStart() {
+        myThread = Thread.currentThread();
+        componentWorkerThread = new NucleusWorkerThread();
+        setComponentWorkerThread(componentWorkerThread);
+    }
+
+    @Override
+    protected void onServiceStop() {
+        if (EXIT_CODE_SUCCESS == exitCode || EXIT_CODE_TERMINATED == exitCode
+                || REQUEST_RESTART == exitCode || REQUEST_REBOOT == exitCode) {
+            errorDetected = false;
+        }
+    }
+
+    /**
+     * Resets start attempts counter back to zero.
+     */
+    public static void resetStartAttemptsCounter() {
+        synchronized (startAttemptsCounter) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Object startAttemptsCounter synchronized on line 190 should not be re-assigned in the same synchronized block on line 191. After this re-assignment, you lose the reference to the synchronized object.